### PR TITLE
Refactor kubernetes client and fix unit tests

### PR DESCRIPTION
**Pull request by chandankumar4**
- Refactor kubernetes client
- Enable unit test for watch repo
- Add test cases for apply and delete resource

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -17,6 +17,8 @@ jobs:
         with:
           go-version: '1.20'
         id: go
+      - name: Install MockGen
+        run:  go install github.com/golang/mock/mockgen@v1.6.0
```
Comment (chandankumar4) [inline .github/workflows/ci.yaml:21]: Installing `mockgen` in pipeline for generating the mock client code

File: Makefile [hunk 2]
```diff
@@ -72,7 +73,7 @@ vet: ## Run go vet against code.
 
 .PHONY: test
 test: manifests generate fmt vet envtest ## Run tests.
-	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) --bin-dir $(LOCALBIN) -p path)" go test ./... -coverprofile cover.out
+	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) --bin-dir $(LOCALBIN) -p path)" go test -v ./... -coverprofile cover.out
```
Comment (chandankumar4) [inline Makefile:76]: Added `-v` for extra verbose while running it, which can help to debug tests easiliy.

File: internal/kubernetes/interface.go [hunk 3]
```diff
@@ -1,14 +1,29 @@
...
 import (
+	"context"
+
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8sClient "sigs.k8s.io/controller-runtime/pkg/client"
 )
 
+// Generate a new client using the kubernetes controller.
+
+//go:generate mockgen -destination=mocks/mock_client.go -package=mocks . Client
```
Comment (chandankumar4) [inline internal/kubernetes/interface.go:12]: Added comment for generating the mock client

**Review by xdevxy — APPROVED** [review]
LGTM, thanks a lot @chandankumar4!

File: tests/utils/utils.go [hunk 4]
```diff
@@ -2,15 +2,22 @@ package utils
...
-	data := []byte(`apiVersion: apps/v1
+	obj := make(map[string]interface{})
+	if err := yaml.Unmarshal(GetTestManifest(), obj); err != nil {
+		log.Fatalf("failed to unmarshal yaml data, err: %v", err)
+	}
+
+	return obj
+}
+
+func GetTestManifest() []byte {
```
Comment (juliev0) [inline tests/utils/utils.go:19]: thanks for breaking this into multiple functions

**Review by juliev0 — APPROVED** [review]

File: internal/kubernetes/client.go [hunk 5]
```diff
@@ -170,3 +169,18 @@ func SetNamespaceIfScoped(namespaceOverride string, u *unstructured.Unstructured
...
+
+// Update the resource in kubernetes cluster
+func (c *client) Update(ctx context.Context, obj k8sClient.Object, opts ...k8sClient.UpdateOption) error {
+	return c.kubeClient.Update(ctx, obj, opts...)
+}
+
+// StatusUpdate will update the status of kubernetes resource
+func (c *client) StatusUpdate(ctx context.Context, obj k8sClient.Object, opts ...k8sClient.SubResourceUpdateOption) error {
+	return c.kubeClient.Status().Update(ctx, obj, opts...)
+}
```
Comment (juliev0) [inline internal/kubernetes/client.go:186]: love these generic methods used by both the Reconciler and the apply code!

**Pull request merged**